### PR TITLE
Build needed cross toolchain with buildroot for non-x86_64 hosts

### PR DESCRIPTION
**Pull request by jenswi-linaro**
The needed precompiled toolchains provided by Arm are only available for
x86_64 hosts. For all other host architectures we build it with
Buildroot itself.

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

### CODE REVIEW

**Review by jforissier — CHANGES_REQUESTED** [review]
That looks like a smart way to support building on Arm systems, right? :wink: I'll give it a try.

File: toolchain.mk [hunk 1]
```diff
@@ -4,7 +4,40 @@
 SHELL				= /bin/bash
 ROOT				?= $(CURDIR)/..
 TOOLCHAIN_ROOT 			?= $(ROOT)/toolchains
+UNAME_M				:= $(shell uname -m)
 
+ifneq ($(UNAME_M),x86_32)
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply toolchain.mk:9]: Makes sense, I'll fix.

File: common.mk [hunk 2]
```diff
@@ -240,12 +245,16 @@ BUILDROOT_TOOLCHAIN=toolchain-br # Use toolchain supplied by buildroot
 DEFCONFIG_GDBSERVER=--br-defconfig build/br-ext/configs/gdbserver.conf
 else
 # Local toolchains (downloaded by "make toolchains")
+ifeq ($(UNAME_M),aarch64)
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply common.mk:248]: Missed this one, I'll fix.

**Comment by jenswi-linaro** [comment]
> That looks like a smart way to support building on Arm systems, right?  I'll give it a try.

Yes, it was surprisingly easy once I decided on this course. If others do like me and have a toolchains directory shared between all repos via a symlink you don't need to recompile often enough for this to be too inconvenient.

**Comment by jenswi-linaro** [comment]
Update

File: br-ext/configs/toolchain-aarch32-sdk [hunk 3]
```diff
@@ -0,0 +1,9 @@
+BR2_TOOLCHAIN_EXTERNAL=y
+BR2_TOOLCHAIN_EXTERNAL_CUSTOM=y
+BR2_TOOLCHAIN_EXTERNAL_PATH="%TOP_DIR%/toolchains/aarch64"
```
Comment (etienne-lms) [inline br-ext/configs/toolchain-aarch32-sdk:3]: s/aarch64/aarch32/

Comment (jenswi-linaro) [reply br-ext/configs/toolchain-aarch32-sdk]: Thanks, I'll fix.

**Comment by jenswi-linaro** [comment]
Update

File: br-ext/configs/toolchain-aarch32-sdk [hunk 4]
```diff
@@ -0,0 +1,9 @@
+BR2_TOOLCHAIN_EXTERNAL=y
+BR2_TOOLCHAIN_EXTERNAL_CUSTOM=y
+BR2_TOOLCHAIN_EXTERNAL_PATH="%TOP_DIR%/toolchains/aarch32"
+BR2_TOOLCHAIN_EXTERNAL_CUSTOM_PREFIX="arm-linux"
```
Comment (jforissier) [inline br-ext/configs/toolchain-aarch32-sdk:4]: Why not use the traditional prefix, `arm-linux-gnueabihf`? Wouldn't it simplify things a bit?

File: br-ext/configs/toolchain-aarch64-sdk [hunk 5]
```diff
@@ -0,0 +1,9 @@
+BR2_TOOLCHAIN_EXTERNAL=y
+BR2_TOOLCHAIN_EXTERNAL_CUSTOM=y
+BR2_TOOLCHAIN_EXTERNAL_PATH="%TOP_DIR%/toolchains/aarch64"
+BR2_TOOLCHAIN_EXTERNAL_CUSTOM_PREFIX="aarch64-linux"
```
Comment (jforissier) [inline br-ext/configs/toolchain-aarch64-sdk:4]: Why not keep the traditional prefix, `aarch64-linux-gnu`?

File: toolchain.mk [hunk 6]
```diff
@@ -4,7 +4,9 @@
 SHELL				= /bin/bash
 ROOT				?= $(CURDIR)/..
 TOOLCHAIN_ROOT 			?= $(ROOT)/toolchains
+UNAME_M				:= $(shell uname -m)
 
+ifeq ($(UNAME_M),x86_32)
```
Comment (jforissier) [inline toolchain.mk:9]: `x86_64`?

Comment (jenswi-linaro) [reply toolchain.mk:9]: Whoops, I'll fix.

File: br-ext/configs/toolchain-aarch32-sdk (hunk 4 above)
Comment (jenswi-linaro) [reply br-ext/configs/toolchain-aarch32-sdk:4]: I don't know how to configure "make sdk" to do that. Here we're using what "make sdk" produced.

File: br-ext/configs/toolchain-aarch64-sdk (hunk 5 above)
Comment (jenswi-linaro) [reply br-ext/configs/toolchain-aarch64-sdk:4]: I don't know how to configure "make sdk" to do that. Here we're using what "make sdk" produced.

Comment (jforissier) [reply br-ext/configs/toolchain-aarch64-sdk:4]: Ah, that's fine then.

**Comment by jbech-linaro** [comment]
I've just glimpsed the changes, but does this require updates to the "build" section in our docs?

**Comment by jenswi-linaro** [comment]
> I've just glimpsed the changes, but does this require updates to the "build" section in our docs?

Not really, unless we'd like to be very clear on what's going on.

**Comment by jbech-linaro** [comment]
> > I've just glimpsed the changes, but does this require updates to the "build" section in our docs?
> 
> Not really, unless we'd like to be very clear on what's going on.

OK, thanks for confirming.

File: toolchain.mk [hunk 7]
```diff
@@ -55,3 +57,35 @@ endef
 .PHONY: clang-toolchains
 clang-toolchains:
 	$(call dl-clang,$(CLANG_VER),$(CLANG_PATH))
+
+else # $(UNAME_M) != x86_32
```
Comment (jforissier) [inline toolchain.mk:61]: `!= x86_64` :wink:

**Review by jforissier — APPROVED** [review]
One more small fix needed... With that:

`Reviewed-by: Jerome Forissier <jerome@forissier.org>`
`Tested-by: Jerome Forissier <jerome@forissier.org> (uname -m: aarch64)`

File: toolchain.mk (hunk 7 above)
Comment (jenswi-linaro) [reply toolchain.mk:61]: Thanks, I'll fix

**Comment by jenswi-linaro** [comment]
Squashed and tags applied

File: common.mk [hunk 8]
```diff
@@ -103,8 +104,12 @@ QEMU_VIRTFS_MOUNTPOINT	?= /mnt/host
 # Mandatory for autotools (for specifying --host)
 ################################################################################
 ifeq ($(COMPILE_NS_USER),64)
+ifeq ($(UNAME_M),x86_64)
 MULTIARCH			:= aarch64-linux-gnu
 else
+MULTIARCH			:= aarch64-linux
+endif
+else
 MULTIARCH			:= arm-linux-gnueabihf
```
Comment (jforissier) [inline common.mk:113]: I have a doubt about this, given that 
`AARCH32_CROSS_COMPILE := $(AARCH32_PATH)/bin/arm-linux-` below when `!x86_64`. It looks like `MULTIARCH` should be `arm-linux` in this case. Unfortunately I can't test at the moment.

Comment (jenswi-linaro) [reply common.mk:113]: I'm testing. It seems that something needs to be changed at least, stay tuned.

**Comment by jenswi-linaro** [comment]
Got it working for AArch32, please take a look at the fix.

**Comment by jforissier** [comment]
Looks good :+1:  My R-b still applies and please update my T-b like so:
```
Tested-by: Jerome Forissier <jerome@forissier.org> (QEMUv8 64/64, host uname -m: aarch64)
```

**Comment by jforissier** [comment]
IBART was updated to GCC 10.2 due to https://github.com/OP-TEE/build/pull/491 being merged, I guess this means all references to GCC 9 need to be updated to 10 here also.

**Comment by jenswi-linaro** [comment]
> IBART was updated to GCC 10.2 due to #491 being merged, I guess this means all references to GCC 9 need to be updated to 10 here also.

OK, I'm rebasing and updating.

**Comment by jenswi-linaro** [comment]
Rebased and updated with an upgrade commit.

**Comment by jforissier** [comment]
My previous T-b still applies, and in addition COMPILE_S_USER=32 is OK too:
```
Tested-by: Jerome Forissier <jerome@forissier.org> (QEMUv8 64/32, host uname -m: aarch64)
```

**Comment by jforissier** [comment]
Will you squash the two commits perhaps? I don't think the first one works on it own, does it?

**Comment by jenswi-linaro** [comment]
> Will you squash the two commits perhaps? I don't think the first one works on it own, does it?

That's not necessary, this toolchain is independent of the downloaded one.

**Comment by jforissier** [comment]
> > Will you squash the two commits perhaps? I don't think the first one works on it own, does it?
> 
> That's not necessary, this toolchain is independent of the downloaded one.

Hmmmmmm then why did IBART fail here: https://optee.mooo.com:5000/logs/OP-TEE/build/488/663702791/88301d630c6c7637bb599c9a7edc6f6457fcd017#end-01-hikey6220-build?

**Comment by jbech-linaro** [comment]
> > > Will you squash the two commits perhaps? I don't think the first one works on it own, does it?
> > 
> > 
> > That's not necessary, this toolchain is independent of the downloaded one.
> 
> Hmmmmmm then why did IBART fail here: https://optee.mooo.com:5000/logs/OP-TEE/build/488/663702791/88301d630c6c7637bb599c9a7edc6f6457fcd017#end-01-hikey6220-build?

Just beware that IBART has two aarch32/64 gcc pre-downloaded and they are kept outside the forest. I.e., for every job we start with a `rm -rf *` and symlink the two gcc toolchains. If this PR introduced caching that are kept in tree, then I probably need to make changes to IBART as well.

**Comment by jenswi-linaro** [comment]
> > > Will you squash the two commits perhaps? I don't think the first one works on it own, does it?
> > 
> > 
> > That's not necessary, this toolchain is independent of the downloaded one.
> 
> Hmmmmmm then why did IBART fail here: https://optee.mooo.com:5000/logs/OP-TEE/build/488/663702791/88301d630c6c7637bb599c9a7edc6f6457fcd017#end-01-hikey6220-build?

Probably because I hadn't rebased the branch.

**Comment by jenswi-linaro** [comment]
> Just beware that IBART has two aarch32/64 gcc pre-downloaded and they are kept outside the forest. I.e., for every job we start with a `rm -rf *` and symlink the two gcc toolchains. If this PR introduced caching that are kept in tree, then I probably need to make changes to IBART as well.

IBART is an x86_64 host so it shouldn't be affected by this.

**Comment by jforissier** [comment]
> > > > Will you squash the two commits perhaps? I don't think the first one works on it own, does it?
> > > 
> > > 
> > > That's not necessary, this toolchain is independent of the downloaded one.
> > 
> > 
> > Hmmmmmm then why did IBART fail here: https://optee.mooo.com:5000/logs/OP-TEE/build/488/663702791/88301d630c6c7637bb599c9a7edc6f6457fcd017#end-01-hikey6220-build?
> 
> Probably because I hadn't rebased the branch.

Totally makes sense. You hadn't rebased but the pre-downloaded toolchain was 10.2 already. Thanks for clarifying, I'm waiting for IBART and I'll merge this. Again, nice feature :)

**Pull request merged**